File: server/formation_test.go
package main

import (
	"math"
	"testing"
)

func TestLaunchFormationAndRestarts(t *testing.T) {
	s := initial()
	expected := [18][2]int{{320, 1104}, {213, 992}, {426, 992}, {213, 768}, {426, 768}, {320, 800}, {106, 576}, {426, 576}, {320, 640}, {320, 48}, {426, 160}, {213, 160}, {426, 384}, {213, 384}, {320, 352}, {533, 576}, {213, 576}, {320, 512}}
	check := func(swapped bool) {
		t.Helper()
		for i, p := range s.Players {
			j := i
			if swapped {
				j = (i + 9) % 18
			}
			got := [2]int{int(math.Round(p.Z/(22.4/576) + 320)), int(math.Round(576 - p.X/(22.4/576)))}
			if got != expected[j] {
				t.Fatalf("player %d: %v, want %v", i, got, expected[j])
			}
		}
	}
	check(false)
	s.selectPlayers()
	if s.Controlled != [2]int{8, 17} {
		t.Fatal(s.Controlled)
	}
	s.Tick = 1
	s.Players[3].Health = 63
	s.Players[3].Stats[0] = 180
	for i := range s.Players {
		s.Players[i].X = 0
		s.Players[i].Z = 0
	}
	s.resetPitch()
	check(false)
	s.Period = 2
	s.resetPitch()
	check(true)
	if s.Players[3].Health != 63 || s.Players[3].Stats[0] != 180 || s.Players[0].FX != -1 || s.Players[9].FX != 1 {
		t.Fatal("restart lost player state or orientation")
	}
}

File: src/formation.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, resetPitch, selectPlayers } from "./game.js";

test("launch formation, goal restart and halftime side exchange", () => {
  const s = initial();
  const unit = 22.4 / 576;
  const terrain = (p) => [
    Math.round(p.z / unit + 320),
    Math.round(576 - p.x / unit),
  ];
  assert.deepEqual(s.players.slice(0, 9).map(terrain), [
    [320, 1104],
    [213, 992],
    [426, 992],
    [213, 768],
    [426, 768],
    [320, 800],
    [106, 576],
    [426, 576],
    [320, 640],
  ]);
  assert.deepEqual(s.players.slice(9).map(terrain), [
    [320, 48],
    [426, 160],
    [213, 160],
    [426, 384],
    [213, 384],
    [320, 352],
    [533, 576],
    [213, 576],
    [320, 512],
  ]);
  selectPlayers(s);
  assert.deepEqual(s.controlled, [8, 17]);
  const launch = s.players.map(terrain);
  s.tick = 1;
  s.players[3].health = 63;
  s.players[3].stats[0] = 180;
  for (const p of s.players) {
    p.x = 0;
    p.z = 0;
  }
  resetPitch(s);
  assert.deepEqual(s.players.map(terrain), launch);
  assert.equal(s.players[3].health, 63);
  assert.equal(s.players[3].stats[0], 180);
  s.period = 2;
  resetPitch(s);
  assert.deepEqual(s.players.map(terrain), [
    ...launch.slice(9),
    ...launch.slice(0, 9),
  ]);
  assert.equal(s.players[3].health, 63);
  assert.equal(s.players[0].fx, -1);
  assert.equal(s.players[9].fx, 1);
});
